### PR TITLE
Initial implementation of tracing interceptor

### PR DESCRIPTION
**Pull request by pavolloffay**
Initial implementation of tracing interceptor.

- [x] tracing interceptor
- [x] javadoc
- [x] integration tests
- [x] mvnw
- [x] release configuration

### CODE REVIEW

**Comment by pavolloffay** [comment]
Decorator `STANDARD_TAGS` adds HTTP and PEER tags. However on redirects `HTTP_URL` is set to original URL, but PEER tags point to redirected host. 

I think it can be confusing. 

@yurishkuro @bensigelman

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                for (SpanDecorator spanDecorator: decorators) {
+                    spanDecorator.onError(ex, span);
+                }
+                throw ex;
+            } finally {
+                span.finish();
+            }
+        } else {
+            response = chain.proceed(chain.request());
+            activeConnection.set(chain.connection());
```
Comment (bhs) [inline opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java]: I'm confused about how activeConnection.set / get work here. So we're guaranteed that `intercept` will be invoked with `chain.connection()==null` for each HTTP call? And we are also guaranteed that it's on the same thread as this other invocation higher in the stack?

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        Response response;
+
+        // application interceptor?
+        if (chain.connection() == null) {
+            Tracer.SpanBuilder spanBuilder = tracer.buildSpan(chain.request().method())
+                    .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT);
+
+            Object tag = chain.request().tag();
```
Comment (bhs) [inline opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java:76]: so there's only one "tag" allowed per request? I read the okhttp docs, and they make this sound more like a cancellation mechanism (or something)...

**Review by bhs — COMMENTED** [review]
A few initial thoughts.

Overall, I am more and more convinced that https://github.com/opentracing/specification/issues/23 needs formal support from opentracing libraries.

**Comment by pavolloffay** [comment]
@bensigelman thanks for review :+1: 

Basically in all these java-instrumentations users have to pass/access context manually in various ways.  
It would be great to have a standard way to access active/last span context.
It is my first step to instrument various frameworks and then work on how to simplify accessing active span context.

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java (hunk 2 above)
Comment (swankjesse) [reply opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java:76]: It's just a place to stash an object. You can use it however you like. If you need multiple tags, create a `MyTag` class with everything you need.

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java (hunk 1 above)
Comment (swankjesse) [reply opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java]: There's no guarantee that application and network interceptors run on the same thread for the same call.

That said, that's how it works today and there's no particular plans to change that.

**Review by swankjesse — COMMENTED** [review]
Neat. See also https://github.com/square/okhttp/pull/3149

**Comment by bhs** [comment]
@pavolloffay let me know when/if you'd like me to take another look at this. (I'm doing a survey of PRs I should be looking at right now :)

**Comment by pavolloffay** [comment]
@bensigelman thanks! This needs some hacks,  I will let you know :+1:

**Comment by pavolloffay** [comment]
PR updated. I got rid of threadlocal and everything is passed via `request#tag()`. Tag is wrapped to `TagWrapper` with tracing data.

@swankjesse thanks for the review I really appreciate it :+1:  If you could have a quick look it would be great :100: .

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            requestBuilder.tag(new TagWrapper(tagWrapper, span));
+
+            try {
+                response = chain.proceed(requestBuilder.build());
+
+                for (SpanDecorator spanDecorator: decorators) {
+                    spanDecorator.onResponse(response, span);
+                }
+            } catch (Throwable ex) {
+                for (SpanDecorator spanDecorator: decorators) {
```
Comment (swankjesse) [inline opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java:105]: not sure how opentracing works, but I think it’s weird that spans get onErrors for other spans

**Review by swankjesse — COMMENTED** [review]
Looks good

File: opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/SpanDecorator.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            Tags.ERROR.set(span, Boolean.TRUE);
+            span.log(errorLogs(throwable));
+        }
+
+        @Override
+        public void onNetworkResponse(Connection connection, Response response, Span span) {
+            if (response.isRedirect()) {
+                Map<String, String> redirectLogs = new HashMap<>(4);
+                redirectLogs.put("event", "redirect");
+                redirectLogs.put(Tags.PEER_HOSTNAME.getKey(), connection.socket().getInetAddress().getHostName());
```
Comment (objectiser) [inline opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/SpanDecorator.java]: Although the log events can provide a trail to help follow potentially multiple redirections, wondering whether the span also needs these tags for each new peer? i.e. if the opentracing backend does not support logs, then the span data would be misleading. If the span includes multiple peer details, then atleast it shows redirection must have occurred, although won't necessary indicate which was the ultimate peer.

File: opentracing-okhttp3/src/test/java/io/opentracing/contrib/okhttp3/TracingInterceptorTest.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+            okHttpClient.newCall(request).execute();
+        }
+
+        List<MockSpan> mockSpans = mockTracer.finishedSpans();
+        Assert.assertEquals(1, mockSpans.size());
+
+        MockSpan mockSpan = mockSpans.get(0);
+        Assert.assertEquals(200, mockSpan.tags().get(Tags.HTTP_STATUS.getKey()));
+        Assert.assertEquals(mockWebServer.url("foo").toString(), mockSpan.tags().get(Tags.HTTP_URL.getKey()));
+        Assert.assertEquals(1, mockSpan.logEntries().size());
```
Comment (objectiser) [inline opentracing-okhttp3/src/test/java/io/opentracing/contrib/okhttp3/TracingInterceptorTest.java:200]: Is it worth adding checks for the redirect peer details?

**Review by objectiser — APPROVED** [review]
Just some minor comments on the redirect handling, but LGTM :+1:

**Review by bhs — APPROVED** [review]
LGTM, thank you!

**Comment by pavolloffay** [comment]
@bensigelman @swankjesse @objectiser thanks for reviews :+1: !

**Pull request merged**